### DISCUSSION TITLE
Including ts-fresh as feature?

### DISCUSSION THREAD

**Discussion by timonmerk** [neuromodulation/py_neuromodulation#201] [Ideas]
@neumann-wj, @richardkoehler, @lauraflyra check out this time series toolbox: https://tsfresh.readthedocs.io/en/latest/.
Otherall the general concept seems from the feature extraction stage similar to py_neuromodulation. I think it would be a great idea for implementation. So far I realized that it's quite slow, but it could  add potentially hundreds of additional features.

This example shows it's use for the example data, which calculated almost 800 features for a single channel:
https://github.com/neuromodulation/py_neuromodulation/blob/main/examples/example_tsfresh/example_tsfresh.ipynb 

I guess it would be a bit tricky writing a good wrapper around it. But as far as I know this tool seems to be highly used for time series data in other domains, so I guess adding it here would be nice addition.
What do you think?

**Comment by richardkoehler** [comment]
@timonmerk  This toolbox looks super interesting! I think that py_neuromodulation already has a lot of dependencies, and the more dependencies a package has, the more difficult it will be to maintain. So I think it would be worthwhile to think about how many more dependencies will be included in the future :) 
It is already quite hard to use py_neuromodulation for some real-time experiments because both pynm and for example psychopy have a lot of dependencies. I think my ideal solution would be to implement a plug-in architecture for py_neuromodulation. That would give py_neuromodulation the additional advantage of being endlessly "extensible". It shouldn't be that difficult, but it would require some very minor changes to the use of nm_settings.json. I think we already briefly talked about this once during a train ride. This way people could build their custom plug-ins to use additional features if they like. For example, one could create a plug-in called nm-tsfresh, which could then be installed separately, if needed. Let me know what you think! If you are interested in doing this we could also schedule a brief call to discuss this in detail.

**Reply by timonmerk** [reply]
@richardkoehler, thanks for the comment. I completely agree, the installation is currently a struggle. The issue I just saw previously that based on the installation of certain packages libraries need to import differently, e.g. ensure that code still compiles even though a certain library is not installed.
Btw, did you try out py_nm using psychopy at the same time already?

**Comment by richardkoehler** [comment]
@timonmerk I have set up psychopy with py_nm but I haven't tested it in a real-time setting yet!

**Comment by timonmerk** [comment]
@richardkoehler I think I will setup a simple example using `importlib` as explained here https://stackoverflow.com/a/14050282/5060208. Using this method code will hopefully still run even if a package is not installed.

**Comment by timonmerk** [comment]
Another tool that seems relevant for general feature extraction https://github.com/sari-saba-sadiya/EEGExtract
